File: src/main/java/org/capturecoop/cccolorutils/CCHSB.java
package org.capturecoop.cccolorutils;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.capturecoop.ccutils.utils.CCStringUtils;

import java.awt.*;

public class CCHSB {
    private float hue;
    private float saturation;
    private float brightness;
    private int alpha;

    public CCHSB(float hue, float saturation, float brightness) {
        load(hue, saturation, brightness, 255);
    }

    public CCHSB(float hue, float saturation, float brightness, int alpha) {
        load(hue, saturation, brightness, alpha);
    }

    public CCHSB(Color color) {
        float[] values = new float[3];
        Color.RGBtoHSB(color.getRed(), color.getGreen(), color.getBlue(), values);
        load(values[0], values[1], values[2], color.getAlpha());
    }

    private void load(float hue, float saturation, float brightness, int alpha) {
        this.hue = hue;
        this.saturation = saturation;
        this.brightness = brightness;
        this.alpha = alpha;
    }

    public Color toRGB() {
        Color tempColor = new Color(Color.HSBtoRGB(hue, saturation, brightness));
        return new Color(tempColor.getRed(), tempColor.getGreen(), tempColor.getBlue(), alpha);
    }

    public float getHue() {
        return hue;
    }

    public float getSaturation() {
        return saturation;
    }

    public float getBrightness() {
        return brightness;
    }

    public int getAlpha() {
        return alpha;
    }

    @Override
    public String toString() {
        return CCStringUtils.format("CCHSB[hue: %c, saturation: %c, brightness: %c, alpha: %c]", hue, saturation, brightness, alpha);
    }

    @Override
    public boolean equals(Object o) {
        if(this == o)
            return true;

        if(o == null || getClass() != o.getClass())
            return false;

        CCHSB other = (CCHSB) o;

        return new EqualsBuilder()
                .append(hue, other.hue)
                .append(saturation, other.saturation)
                .append(brightness, other.brightness)
                .append(alpha, other.alpha)
                .isEquals();
    }
}

File: src/main/kotlin/org/capturecoop/cccolorutils/CCHSB.kt
package org.capturecoop.cccolorutils

import java.awt.Color

class CCHSB {
    val hue: Float
    val saturation: Float
    val brightness: Float
    val alpha: Int

    constructor(hue: Float, saturation: Float, brightness: Float, alpha: Int = 255) {
        this.hue = hue
        this.saturation = saturation
        this.brightness = brightness
        this.alpha = alpha
    }

    constructor(color: Color) {
        val values = floatArrayOf(0.0F, 0.0F, 0.0F)
        Color.RGBtoHSB(color.red, color.green, color.blue, values)
        this.hue = values[0]
        this.saturation = values[1]
        this.brightness = values[2]
        this.alpha = color.alpha
    }

    fun asColor(): Color {
        Color(Color.HSBtoRGB(hue, saturation, brightness)).also {
            return Color(it.red, it.green, it.blue, alpha)
        }
    }

    override fun toString() = "CCHSB[hue: $hue, saturation: $saturation, brightness: $brightness, alpha: $alpha]"
    override fun equals(other: Any?): Boolean {
        if(other == null || other != this) return false
        if(other !is CCHSB) return false
        return hue == other.hue && saturation == other.saturation && brightness == other.brightness && alpha == other.alpha
    }

    override fun hashCode(): Int {
        var result = hue.hashCode()
        result = 31 * result + saturation.hashCode()
        result = 31 * result + brightness.hashCode()
        result = 31 * result + alpha
        return result
    }

}